malcolm/modules/scanning/parts/scanrunnerpart.py

```diff
@@ -0,0 +1,487 @@
+            "int64", "Number of scans completed",
+            tags=Widget.TEXTUPDATE.tag()
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "scansCompleted",
+            self.scans_completed,
+            self.scans_completed.set_value)
+
+        self.scan_successes = NumberMeta(
+            "int64", "Successful scans",
+            tags=[Widget.TEXTUPDATE.tag()]
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "scanSuccesses",
+            self.scan_successes,
+            self.scan_successes.set_value)
+
+        self.scan_failures = NumberMeta(
+            "int64", "Failed scans",
+            tags=[Widget.TEXTUPDATE.tag()]
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "scanFailures",
+            self.scan_failures,
+            self.scan_failures.set_value)
+
+        self.output_directory = StringMeta(
+            "Root output directory (will create a sub-directory inside)",
+            tags=[config_tag(), Widget.TEXTINPUT.tag()]
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "outputDirectory",
+            self.output_directory,
+            self.output_directory.set_value)
+
+        # Methods
+        registrar.add_method_model(self.loadFile)
+        registrar.add_method_model(self.run, needs_context=True)
+
+    def get_file_contents(self):
+        # type: () -> str
+
+        try:
+            with open(self.scan_file.value, "r") as input_file:
+                return input_file.read()
+        except IOError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value("Could not read scan file")
+            raise
+
+    def parse_yaml(self, string):
+        # type: (str) -> ...
+        try:
+            parsed_yaml = yaml.safe_load(string)
+            return parsed_yaml
+        except yaml.YAMLError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value("Could not parse scan file")
+            raise
+
+    @staticmethod
+    def get_kwargs_from_dict(input_dict, kwargs_list):
+        kwargs = {}
+        if not isinstance(kwargs_list, list):
+            kwargs_list = [kwargs_list]
+        for kwarg in kwargs_list:
+            if kwarg in input_dict:
+                kwargs[kwarg] = input_dict[kwarg]
+        return kwargs
+
+    def parse_compound_generator(self, entry):
+        # type: (dict) -> CompoundGenerator
+        duration = entry['duration']
+        generators = []
+        generators_dict = entry['generators']
+        for generator in generators_dict:
+            generators.append(LineGenerator.from_dict(generator['line']))
+
+        kwargs_list = ['continuous', 'delay_after']
+        kwargs = self.get_kwargs_from_dict(entry, kwargs_list)
+
+        return CompoundGenerator(generators, duration=duration, **kwargs)
```

Comment:
You should be able to do CompoundGenerator.from_dict(entry) here too, as long you have set entry["generators"] = generators

Reply:
I have made this change and added a check for a negative generator duration.